### DISCUSSION TITLE
Integer enum in SQLite3

### DISCUSSION THREAD

**Discussion by keepsimple1** [diesel-rs/diesel#2776] [Q&A] [answered]
Hi,

I am using diesel with SQLite3. What is the best way to support integer enum values in a column? 

For example, I have defined a column like this in SQL schema:
```
    foo_state INTEGER NOT NULL, -- the value will be from an Enum.
```
in `diesel_schema.rs`, I also simply define it as `integer`:
```
        foo_state -> Integer,
```

But in actual source code, I would like to define the struct for inserting and querying including this field as:

```rust
struct Foo {
    pub foo_state: FooState,
}
```
where `FooState` is an Enum like this:
```rust
pub enum FooState {
    Unknown = 0,
    Start = 1,
    End = 2,
}
```

The problem is that it is quite difficult to use the enum directly with diesel ORM methods. I had to implement `FromSql` and `ToSql` traits for `FooState`:
```rust
impl<DB> FromSql<Integer, DB> for FooState {
<snip>
}

impl<DB> ToSql<Integer, DB> for FooState {
<snip>
}
```
After that, I can use the enum in `Insertable` struct,  and `QueryableByName` struct,  but it seems I still cannot use the enum in a `Queryable` struct. 

Did I miss something obvious & easier way to use integer enum fields? Is it possible to use the enum field in a struct as `Queryable` ?  

Thanks.

**Comment by weiznich** [comment, marked as answer]
You likely miss one of [these](https://github.com/diesel-rs/diesel/blob/1.4.x/diesel_tests/tests/custom_types.rs#L22-L23) derives.

Otherwise there is [this](https://docs.rs/crate/diesel-derive-enum/0.4.2) third party crate, which provides an opinionated way to implement such mappings.
